File: IdentityX/testing_framework/permissions_test.go
package testing

import (
	"GoAuth/internal/errx"
	"net/http"
	"testing"
)

func testPermissions(t *testing.T, suite *TestSuite) {
	client := suite.NewClient(t)
	user := client.WithCredentials("permissions@mail.com", ValidPassword).
		Register().
		Login().
		CreateProject("permissions testing")

	projectID := user.projectID
	var permissionID string
	t.Run("CreatePermission", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "document",
				"action": "create",
			}).
			Expect(http.StatusCreated).
			HasMessage("Permission Created").
			RequireDataValue()

		spec := map[string]interface{}{
			"id":         StoreString{Into: &permissionID, Matcher: AnyUUID{}},
			"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
			"object":     "document",
			"action":     "create",
			"created_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	t.Run("GetPermissionByID", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.GET("/projects/" + projectID + "/permissions/" + permissionID).
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"id":         StoreString{Into: &permissionID, Matcher: AnyUUID{}},
			"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
			"object":     "document",
			"action":     "create",
			"created_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	var anotherPermissionID string
	t.Run("CreateAnotherPermission", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "event",
				"action": "read",
			}).
			Expect(http.StatusCreated).
			HasMessage("Permission Created").
			RequireDataValue()

		spec := map[string]interface{}{
			"id":         StoreString{Into: &anotherPermissionID, Matcher: AnyUUID{}},
			"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
			"object":     "event",
			"action":     "read",
			"created_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	t.Run("ListProjectPermissions", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.GET("/projects/" + projectID + "/permissions").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":         AsString{Value: anotherPermissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "event",
				"action":     "read",
				"created_at": AnyDate{},
			},
			map[string]interface{}{
				"id":         AsString{Value: permissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "document",
				"action":     "create",
				"created_at": AnyDate{},
			},
		}

		Validate(t, val, spec)
	})

	t.Run("GetPermissionByObject", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.GET("/projects/"+projectID+"/permissions").
			WithQuery("object", "event").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":         AsString{Value: anotherPermissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "event",
				"action":     "read",
				"created_at": AnyDate{},
			},
		}

		Validate(t, val, spec)
	})

	t.Run("GetPermissionByAction", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.GET("/projects/"+projectID+"/permissions").
			WithQuery("action", "create").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":         AsString{Value: permissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "document",
				"action":     "create",
				"created_at": AnyDate{},
			},
		}

		Validate(t, val, spec)
	})

	var createProductPermissionID string
	t.Run("CreateCreateProductPermission", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "product",
				"action": "create",
			}).
			Expect(http.StatusCreated).
			HasMessage("Permission Created").
			RequireDataValue()

		spec := map[string]interface{}{
			"id":         StoreString{Into: &createProductPermissionID, Matcher: AnyUUID{}},
			"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
			"object":     "product",
			"action":     "create",
			"created_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	t.Run("GetPermissionByActionAgain", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.GET("/projects/"+projectID+"/permissions").
			WithQuery("action", "create").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":         AsString{Value: createProductPermissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "product",
				"action":     "create",
				"created_at": AnyDate{},
			},
			map[string]interface{}{
				"id":         AsString{Value: permissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "document",
				"action":     "create",
				"created_at": AnyDate{},
			},
		}

		Validate(t, val, spec)
	})

	var editProductPermissionID string
	t.Run("CreateEditProductPermission", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "product",
				"action": "edit",
			}).
			Expect(http.StatusCreated).
			HasMessage("Permission Created").
			RequireDataValue()

		spec := map[string]interface{}{
			"id":         StoreString{Into: &editProductPermissionID, Matcher: AnyUUID{}},
			"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
			"object":     "product",
			"action":     "edit",
			"created_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	t.Run("GetPermissionByObjectAgain", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.GET("/projects/"+projectID+"/permissions").
			WithQuery("object", "product").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":         AsString{Value: editProductPermissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "product",
				"action":     "edit",
				"created_at": AnyDate{},
			},
			map[string]interface{}{
				"id":         AsString{Value: createProductPermissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "product",
				"action":     "create",
				"created_at": AnyDate{},
			},
		}

		Validate(t, val, spec)
	})

	t.Run("GetPermissionByObjectAndAction", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.GET("/projects/"+projectID+"/permissions").
			WithQuery("object", "product").
			WithQuery("action", "edit").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":         AsString{Value: editProductPermissionID, Matcher: AnyUUID{}},
				"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
				"object":     "product",
				"action":     "edit",
				"created_at": AnyDate{},
			},
		}

		Validate(t, val, spec)
	})

	t.Run("CreateAdminPermission", func(t *testing.T) {
		val := suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "user",
				"action": "delete",
			}).
			Expect(http.StatusCreated).
			HasMessage("Permission Created").
			RequireDataValue()

		spec := map[string]interface{}{
			"id":         AnyUUID{},
			"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
			"object":     "user",
			"action":     "delete",
			"created_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	t.Run("CreateMasterPermission", func(t *testing.T) {
		val := suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "system",
				"action": "admin",
			}).
			Expect(http.StatusCreated).
			HasMessage("Permission Created").
			RequireDataValue()

		spec := map[string]interface{}{
			"id":         AnyUUID{},
			"project_id": AsString{Value: projectID, Matcher: AnyUUID{}},
			"object":     "system",
			"action":     "admin",
			"created_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	t.Run("CreateDuplicatePermission", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "document",
				"action": "create",
			}).
			Expect(http.StatusConflict).
			HasErrID(errx.PERMissionAlreadyExists).
			HasMessage("permission with object(document) and action(create) already exists")
	})

	t.Run("CreatePermissionNoAction", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "document",
				"action": "",
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.RequestValidationError).
			HasMessage("Validation failed").
			TraceContains("action is required")
	})

	t.Run("CreatePermissionNoObject", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "",
				"action": "create",
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.RequestValidationError).
			HasMessage("Validation failed").
			TraceContains("object is required")
	})

	t.Run("CreateEmptyPermission", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/"+projectID+"/permissions").
			WithBody(map[string]interface{}{
				"object": "",
				"action": "",
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.RequestValidationError).
			HasMessage("Validation failed").
			TraceContains("object is required", "action is required")
	})

	t.Run("CreateInvalidObjectPermission", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "event:*",
				"action": "read",
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.PERMissionInvalidObject).
			HasMessage("invalid permission object: (event:*)")
	})

	t.Run("CreateInvalidActionPermission", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "document",
				"action": "attendance:mark",
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.PERMissionInvalidAction).
			HasMessage("invalid permission action: (attendance:mark)")
	})

	t.Run("CreateWildcardObjectPermission", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "*",
				"action": "read",
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.PERMissionInvalidObject).
			HasMessage("invalid permission object: (*)")
	})

	t.Run("CreateWildcardActionPermission", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).POST("/projects/" + projectID + "/permissions").
			WithBody(map[string]interface{}{
				"object": "document",
				"action": "*",
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.PERMissionInvalidAction).
			HasMessage("invalid permission action: (*)")
	})

	t.Run("NotAllowedQueryParam", func(t *testing.T) {
		suite.NewClient(t).WithAuth(user.auth).GET("/projects/"+projectID+"/permissions").
			WithQuery("not-allowed", "should_deny").
			Expect(http.StatusBadRequest).
			HasErrID(errx.RequestUnknownQueryParam).
			HasMessage("unknown query parameter: not-allowed")
	})
}

File: IdentityX/testing_framework/permissions.test.ts
import { beforeAll, describe, test } from "vitest";
import { createClient } from "./helpers/index.js";
import { assertErrID, assertMessage, assertValidationError, shouldFail } from "./helpers/assert.js";
import {
    Validate,
    AnyUUID,
    AnyDate,
    AsString,
    Store,
} from "./helpers/validate.js";
import { ValidPassword } from "./fixtures/auth/testdata.js";

const ErrPermissionAlreadyExists = "0_PERM_0008_D";
const ErrPermissionInvalidObject = "0_PERM_0007_D";
const ErrPermissionInvalidAction = "0_PERM_0006_D";
const ErrRequestValidation       = "0_REQ_0004_D";
const ErrRequestUnknownQueryParam = "0_REQ_0003_D";

describe("Permissions", () => {
    let user: any;
    let projectID: string;
    let permissionID: string;
    let anotherPermissionID: string;
    let createProductPermissionID: string;
    let editProductPermissionID: string;

    beforeAll(async () => {
        await createClient().withCredentials("permissions@mail.com", ValidPassword).register();
        user = await createClient().withCredentials("permissions@mail.com", ValidPassword).login();
        const project = await user.post("/projects", { project_name: "permissions testing", metadata: { env: "test" } });
        projectID = project.id;
    });

    test("CreatePermission", async () => {
        const ref = { current: "" };
        const data = await user.post(`/projects/${projectID}/permissions`, {
            object: "document",
            action: "create",
        });
        Validate(data, {
            id:         Store(ref, AnyUUID),
            project_id: AsString(projectID, AnyUUID),
            object:     "document",
            action:     "create",
            created_at: AnyDate,
        });
        permissionID = ref.current as string;
    });

    test("GetPermissionByID", async () => {
        const data = await user.get(`/projects/${projectID}/permissions/${permissionID}`);
        Validate(data, {
            id:         AsString(permissionID, AnyUUID),
            project_id: AsString(projectID, AnyUUID),
            object:     "document",
            action:     "create",
            created_at: AnyDate,
        });
    });

    test("CreateAnotherPermission", async () => {
        const ref = { current: "" };
        const data = await user.post(`/projects/${projectID}/permissions`, {
            object: "event",
            action: "read",
        });
        Validate(data, {
            id:         Store(ref, AnyUUID),
            project_id: AsString(projectID, AnyUUID),
            object:     "event",
            action:     "read",
            created_at: AnyDate,
        });
        anotherPermissionID = ref.current as string;
    });

    test("ListProjectPermissions", async () => {
        const data = await user.get(`/projects/${projectID}/permissions`);
        Validate(data, [
            { id: AsString(anotherPermissionID, AnyUUID), project_id: AsString(projectID, AnyUUID), object: "event",    action: "read",   created_at: AnyDate },
            { id: AsString(permissionID,        AnyUUID), project_id: AsString(projectID, AnyUUID), object: "document", action: "create", created_at: AnyDate },
        ]);
    });

    test("GetPermissionByObject", async () => {
        const data = await user.get(`/projects/${projectID}/permissions?object=event`);
        Validate(data, [
            { id: AsString(anotherPermissionID, AnyUUID), project_id: AsString(projectID, AnyUUID), object: "event", action: "read", created_at: AnyDate },
        ]);
    });

    test("GetPermissionByAction", async () => {
        const data = await user.get(`/projects/${projectID}/permissions?action=create`);
        Validate(data, [
            { id: AsString(permissionID, AnyUUID), project_id: AsString(projectID, AnyUUID), object: "document", action: "create", created_at: AnyDate },
        ]);
    });

    test("CreateCreateProductPermission", async () => {
        const ref = { current: "" };
        const data = await user.post(`/projects/${projectID}/permissions`, {
            object: "product",
            action: "create",
        });
        Validate(data, {
            id:         Store(ref, AnyUUID),
            project_id: AsString(projectID, AnyUUID),
            object:     "product",
            action:     "create",
            created_at: AnyDate,
        });
        createProductPermissionID = ref.current as string;
    });

    test("GetPermissionByActionAgain", async () => {
        const data = await user.get(`/projects/${projectID}/permissions?action=create`);
        Validate(data, [
            { id: AsString(createProductPermissionID, AnyUUID), project_id: AsString(projectID, AnyUUID), object: "product",  action: "create", created_at: AnyDate },
            { id: AsString(permissionID,              AnyUUID), project_id: AsString(projectID, AnyUUID), object: "document", action: "create", created_at: AnyDate },
        ]);
    });

    test("CreateEditProductPermission", async () => {
        const ref = { current: "" };
        const data = await user.post(`/projects/${projectID}/permissions`, {
            object: "product",
            action: "edit",
        });
        Validate(data, {
            id:         Store(ref, AnyUUID),
            project_id: AsString(projectID, AnyUUID),
            object:     "product",
            action:     "edit",
            created_at: AnyDate,
        });
        editProductPermissionID = ref.current as string;
    });

    test("GetPermissionByObjectAgain", async () => {
        const data = await user.get(`/projects/${projectID}/permissions?object=product`);
        Validate(data, [
            { id: AsString(editProductPermissionID,   AnyUUID), project_id: AsString(projectID, AnyUUID), object: "product", action: "edit",   created_at: AnyDate },
            { id: AsString(createProductPermissionID, AnyUUID), project_id: AsString(projectID, AnyUUID), object: "product", action: "create", created_at: AnyDate },
        ]);
    });

    test("GetPermissionByObjectAndAction", async () => {
        const data = await user.get(`/projects/${projectID}/permissions?object=product&action=edit`);
        Validate(data, [
            { id: AsString(editProductPermissionID, AnyUUID), project_id: AsString(projectID, AnyUUID), object: "product", action: "edit", created_at: AnyDate },
        ]);
    });

    test("CreateAdminPermission", async () => {
        const data = await user.post(`/projects/${projectID}/permissions`, {
            object: "user",
            action: "delete",
        });
        Validate(data, {
            id:         AnyUUID,
            project_id: AsString(projectID, AnyUUID),
            object:     "user",
            action:     "delete",
            created_at: AnyDate,
        });
    });

    test("CreateMasterPermission", async () => {
        const data = await user.post(`/projects/${projectID}/permissions`, {
            object: "system",
            action: "admin",
        });
        Validate(data, {
            id:         AnyUUID,
            project_id: AsString(projectID, AnyUUID),
            object:     "system",
            action:     "admin",
            created_at: AnyDate,
        });
    });

    test("CreateDuplicatePermission", async () => {
        const body = await shouldFail(
            user.post(`/projects/${projectID}/permissions`, { object: "document", action: "create" }),
            409
        );
        assertErrID(body, ErrPermissionAlreadyExists);
        assertMessage(body, "permission with object(document) and action(create) already exists");
    });

    test("CreatePermissionNoAction", async () => {
        const body = await shouldFail(
            user.post(`/projects/${projectID}/permissions`, { object: "document", action: "" }),
            400
        );
        assertErrID(body, ErrRequestValidation);
        assertMessage(body, "Validation failed");
        assertValidationError(body, "action is required");
    });

    test("CreatePermissionNoObject", async () => {
        const body = await shouldFail(
            user.post(`/projects/${projectID}/permissions`, { object: "", action: "create" }),
            400
        );
        assertErrID(body, ErrRequestValidation);
        assertMessage(body, "Validation failed");
        assertValidationError(body, "object is required");
    });

    test("CreateEmptyPermission", async () => {
        const body = await shouldFail(
            user.post(`/projects/${projectID}/permissions`, { object: "", action: "" }),
            400
        );
        assertErrID(body, ErrRequestValidation);
        assertMessage(body, "Validation failed");
        assertValidationError(body, "object is required", "action is required");
    });

    test("CreateInvalidObjectPermission", async () => {
        const body = await shouldFail(
            user.post(`/projects/${projectID}/permissions`, { object: "event:*", action: "read" }),
            400
        );
        assertErrID(body, ErrPermissionInvalidObject);
        assertMessage(body, "invalid permission object: (event:*)");
    });

    test("CreateInvalidActionPermission", async () => {
        const body = await shouldFail(
            user.post(`/projects/${projectID}/permissions`, { object: "document", action: "attendance:mark" }),
            400
        );
        assertErrID(body, ErrPermissionInvalidAction);
        assertMessage(body, "invalid permission action: (attendance:mark)");
    });

    test("CreateWildcardObjectPermission", async () => {
        const data = await user.post(`/projects/${projectID}/permissions`, { object: "*", action: "read" });
        Validate(data, {
            id:         AnyUUID,
            project_id: AsString(projectID, AnyUUID),
            object:     "*",
            action:     "read",
            created_at: AnyDate,
        });
    });

    test("CreateWildcardActionPermission", async () => {
        const data = await user.post(`/projects/${projectID}/permissions`, { object: "document", action: "*" });
        Validate(data, {
            id:         AnyUUID,
            project_id: AsString(projectID, AnyUUID),
            object:     "document",
            action:     "*",
            created_at: AnyDate,
        });
    });

    test("CreateFullWildcardPermission", async () => {
        const data = await user.post(`/projects/${projectID}/permissions`, { object: "*", action: "*" });
        Validate(data, {
            id:         AnyUUID,
            project_id: AsString(projectID, AnyUUID),
            object:     "*",
            action:     "*",
            created_at: AnyDate,
        });
    });

    test("NotAllowedQueryParam", async () => {
        const body = await shouldFail(
            user.get(`/projects/${projectID}/permissions?not-allowed=should_deny`),
            400
        );
        assertErrID(body, ErrRequestUnknownQueryParam);
        assertMessage(body, "unknown query parameter: not-allowed");
    });
});